### PR TITLE
Refactor lagrangian function

### PR DESCRIPTION
**Pull request by tmigot**
This would help unify NLS and NLP models for the usual NLP API.

Connected to #111 .

### CODE REVIEW

File: src/ADNLPModels.jl [hunk 1]
```diff
@@ -56,6 +56,40 @@ function get_F(nls::AbstractADNLSModel)
...
+function get_lag(nlp::AbstractADNLPModel, ::ADBackend, obj_weight::Real)
+  return ℓ(x; obj_weight = obj_weight) = obj_weight * nlp.f(x)
+end
+function get_lag(nlp::AbstractADNLPModel, b::ADBackend, obj_weight::Real, y::AbstractVector)
+  if nlp.meta.nnln == 0
+    return get_lag(nlp, b, obj_weight)
+  end
+  c = get_c(nlp, b)
+  yview = (length(y) == nlp.meta.nnln) ? y : view(y, (nlp.meta.nlin + 1):(nlp.meta.ncon))
+  ℓ(x; obj_weight = obj_weight, y = y) = obj_weight * nlp.f(x) + dot(c(x), yview)
```
Comment (amontoison) [inline src/ADNLPModels.jl:74]:
```suggestion
  ℓ(x; obj_weight = obj_weight, y = yview) = obj_weight * nlp.f(x) + dot(c(x), yview)
```

File: src/ADNLPModels.jl [hunk 2]
```diff
@@ -56,6 +56,40 @@ function get_F(nls::AbstractADNLSModel)
...
+  return ℓ(x; obj_weight = obj_weight) = obj_weight * mapreduce(Fi -> Fi^2, +, F(x)) / 2
+end
+function get_lag(nls::AbstractADNLSModel, b::ADBackend, obj_weight::Real, y::AbstractVector)
+  if nls.meta.nnln == 0
+    return get_lag(nls, b, obj_weight)
+  end
+  F = get_F(nls, b)
+  c = get_c(nls, b)
+  yview = (length(y) == nls.meta.nnln) ? y : view(y, (nls.meta.nlin + 1):(nls.meta.ncon))
+  ℓ(x; obj_weight = obj_weight, y = y) = obj_weight * sum(F(x) .^ 2) / 2 + dot(c(x), yview)
```
Comment (amontoison) [inline src/ADNLPModels.jl:89]:
```suggestion
  ℓ(x; obj_weight = obj_weight, y = yview) = obj_weight * sum(F(x) .^ 2) / 2 + dot(c(x), yview)
```

File: src/ADNLPModels.jl [hunk 3]
```diff
@@ -56,6 +56,40 @@ function get_F(nls::AbstractADNLSModel)
...
+  end
+  c = get_c(nlp, b)
+  yview = (length(y) == nlp.meta.nnln) ? y : view(y, (nlp.meta.nlin + 1):(nlp.meta.ncon))
+  ℓ(x; obj_weight = obj_weight, y = y) = obj_weight * nlp.f(x) + dot(c(x), yview)
+  return ℓ
+end
+
+function get_lag(nls::AbstractADNLSModel, b::ADBackend, obj_weight::Real)
+  F = get_F(nls, b)
+  return ℓ(x; obj_weight = obj_weight) = obj_weight * mapreduce(Fi -> Fi^2, +, F(x)) / 2
```
Comment (amontoison) [inline src/ADNLPModels.jl:80]:
```suggestion
  ℓ(x; obj_weight = obj_weight) = obj_weight * mapreduce(Fi -> Fi^2, +, F(x)) / 2
  return ℓ
```

File: src/ADNLPModels.jl [hunk 4]
```diff
@@ -56,6 +56,40 @@ function get_F(nls::AbstractADNLSModel)
 end
 get_F(nls::AbstractADNLSModel, ::ADBackend) = get_F(nls)
 
+"""
+    get_lag(nlp, ::ADBackend, obj_weight)
```
Comment (amontoison) [inline src/ADNLPModels.jl:60]:
```suggestion
    get_lag(nlp, b::ADBackend, obj_weight)
```

File: src/ADNLPModels.jl [hunk 5]
```diff
@@ -56,6 +56,40 @@ function get_F(nls::AbstractADNLSModel)
...
+    get_lag(nlp, ::ADBackend, obj_weight, y)
```
Comment (amontoison) [inline src/ADNLPModels.jl:61]:
```suggestion
    get_lag(nlp, b::ADBackend, obj_weight, y)
```

File: src/ADNLPModels.jl [hunk 6]
```diff
@@ -56,6 +56,40 @@ function get_F(nls::AbstractADNLSModel)
...
+
+Return the lagrangian function `ℓ(x) = obj_weight * f(x) + c(x)ᵀy`.
+"""
+function get_lag(nlp::AbstractADNLPModel, ::ADBackend, obj_weight::Real)
```
Comment (amontoison) [inline src/ADNLPModels.jl:65]:
```suggestion
function get_lag(nlp::AbstractADNLPModel, b::ADBackend, obj_weight::Real)
```

File: src/ADNLPModels.jl (hunk 1 above)
Comment (amontoison) [reply src/ADNLPModels.jl:74]: Tangi, we should add a test for this corner case.

File: src/ADNLPModels.jl (hunk 2 above)
Comment (amontoison) [reply src/ADNLPModels.jl:89]: We should add a test here too.

**Review by amontoison — APPROVED** [review]
